fix(nuxt,schema): allow showing spa loader til after hydration

RokeAlvo opened the pull request:
### 🔗 Linked issue

Resolves #21721

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality)
- [] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

Currently, the SPA loader is displayed only until the application is mounted. After that, the user sees a white screen until suspense:resolved occurs.

Changes:
  - Show SPA loading template outside of app root element
  - Add configuration options for SPA loading template element (app.spaLoaderTag, app.spaLoaderAttrs)
  - Add flag to config.experimental.spaPreloaderOutside (default value is false)

### 📝 Checklist


- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
  - Introduced `spaLoadingTemplateLocation` for customizable loading screens in client-only pages.
  - Added `spaLoaderTag` and `spaLoaderAttrs` for enhanced SPA loader configuration.
  - New Vue component `app.vue` with asynchronous data handling.
  - New HTML element for SPA loading indication.

- **Bug Fixes**
  - Improved handling of SPA loader visibility based on configuration settings.

- **Documentation**
  - Updated upgrade guide with instructions for transitioning to Nuxt 4.
  - Enhanced documentation on experimental features, including `spaLoadingTemplateLocation`.

- **Tests**
  - Added tests to validate SPA loader behavior based on configuration settings.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

RokeAlvo committed:
fix(nuxt, schema) Keep showing the spa-loading-template until suspense:resolve (#21721)

danielroe committed:
Merge remote-tracking branch 'origin/main' into fix/21721-spa-loading

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

danielroe reviewed (commented):
This looks like a nice fix to me but will need to defer to v4 as it's might be a breaking change (CSS might be targeted assuming the loading element is within the root app).

Alternatively we could take an experimental option that 'switches on' this behaviour, which is enabled by default in v4 mode, and merge sooner.

RokeAlvo commented:
> Alternatively we could take an experimental option that 'switches on' this behaviour, which is enabled by default in v4 mode, and merge sooner.

add `experimental.spaLoaderOutside` flag to nuxt.config?

RokeAlvo committed:
Merge branch 'main' into fix/21721-spa-loading

danielroe commented:
that makes sense. and you can also use `routeRules` with `ssr: false` to turn on spa loading, even within the basic fixture, which might be a simpler approach...

RokeAlvo committed:
add experimental.spaPreloaderOutside flag

RokeAlvo committed:
add tests for sapPreloaderOutside

RokeAlvo committed:
Merge branch 'main' into fix/21721-spa-loading

RokeAlvo commented:
> that makes sense. and you can also use `routeRules` with `ssr: false` to turn on spa loading, even within the basic fixture, which might be a simpler approach...

- added experimenal flag
- changed tests

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

RokeAlvo committed:
Merge branch 'main' into fix/21721-spa-loading

RokeAlvo committed:
Merge branch 'main' into fix/21721-spa-loading

danielroe committed:
perf: tree-shake spa loader handling

danielroe committed:
Merge remote-tracking branch 'origin/main' into fix/21721-spa-loading

danielroe committed:
chore: initialise strings within spa template

danielroe committed:
fix: set default `spaPreloaderOutside` when `compatibilityVersion: 4`

danielroe committed:
refactor: rename to `spaLoadingTemplateLocation`

danielroe committed:
docs: add documentation

danielroe committed:
chore: remove default value for `spaLoaderAttrs`

danielroe committed:
chore: lint

danielroe committed:
test: use correct test id

danielroe commented on test/spa-loader/spa-preloader-outside-disabled.test.ts:
```diff
@@ -0,0 +1,41 @@
+    expect(html).toContain(`<div id="__nuxt"><div data-testid="loader">loading...</div>\n</div>`)
+  })
+
+  it('spa-loader does not appear while the app is mounting', async () => {
+    const browser = await getBrowser()
+    const page = await browser.newPage({})
+    await page.goto(url('/spa'), { waitUntil: 'domcontentloaded' })
+
+    const loader = page.getByTestId('loader')
+    expect(await loader.isHidden()).toBeTruthy()
+
+    await page.close()
+  }, 60_000)
```
what is the desired behaviour we're testing here?

(or is this just proof of the current bug?)

RokeAlvo replied:
Correct. Since we maintain the current behavior with the value of within, it needs to be established.

danielroe commented on test/spa-loader/spa-preloader-outside-enabled.test.ts:
```diff
@@ -0,0 +1,52 @@
+
+    const content = page.getByTestId('content')
+    await content.waitFor({ state: 'visible' })
+    expect(await loader.isHidden()).toBeTruthy()
```
maybe we need a slightly different assertion? or modify the implementation so both are not present on the page at the same moment?

RokeAlvo replied:
This is a test for the /ssr page. We should not see the SPA loader at any time:

- neither during document loading (line 44)
- nor while resolving promises. (line 48)

RokeAlvo replied:
Hmm... the first test is broken...

RokeAlvo committed:
Merge branch 'main' into fix/21721-spa-loading

RokeAlvo commented on packages/schema/src/config/app.ts:
```diff
@@ -274,14 +274,7 @@ export default defineUntypedSchema({
-      $resolve: async (val: undefined | null | Record<string, unknown>, get) => {
-        const spaLoaderId = await get('app.spaLoaderId')
-        return defu(val, {
-          id: spaLoaderId === false ? undefined : (spaLoaderId || '__nuxt-spa-loader'),
```
@danielroe We need to ensure that the element's ID is set. These changes break the preloader.

danielroe replied:
ahh, of course. I was just intending to get rid of the reference to the nonexistent spaPreloaderId.

danielroe replied:
... and what do you mean it breaks the preloader?

danielroe committed:
chore: add back default id value for spaLoaderAttrs

This partially reverts commit 6ca7600a57aa657f9b1c18142cce27f1c64f67c0.

RokeAlvo commented on packages/schema/src/config/app.ts:
```diff
@@ -274,7 +274,9 @@ export default defineUntypedSchema({
      * @type {typeof import('@unhead/schema').HtmlAttributes}
      */
-    spaLoaderAttrs: {},
+    spaLoaderAttrs: {
```
maybe, better?
```
    spaLoaderAttrs: {
<<<<<<< HEAD
      $resolve: (val: undefined | null | Record<string, unknown>) => {
        return defu(val, {
          id: val?.id ?? '__nuxt-spa-loader',
        })
      },
=======
      id: '__nuxt-loader',
>>>>>>> f4106a79061f41753ee9caff028d8755272634af
    },
  },
```

danielroe replied:
not sure I understand

RokeAlvo replied:
So, if I pass

spaLoaderAttrs: { class: "some-class" }
it will ultimately result in

{ class: "some-class", id: "__nuxt-loader" }
right?

RokeAlvo committed:
skip test in dev mode

RokeAlvo commented on test/spa-loader/spa-preloader-outside-disabled.test.ts:
I don't know how to fix it.

RokeAlvo committed:
Merge branch 'main' into fix/21721-spa-loading

danielroe committed:
Merge remote-tracking branch 'origin/main' into fix/21721-spa-loading

danielroe committed:
test: remove server dir

danielroe committed:
chore: update workspace name

danielroe committed:
chore: remove comment

danielroe committed:
tetst: remove newline

danielroe committed:
fix: global re

danielroe committed:
chore: returns too

danielroe committed:
Merge remote-tracking branch 'origin/main' into fix/21721-spa-loading

danielroe committed:
chore: add engines

danielroe committed:
test: update bundle size
